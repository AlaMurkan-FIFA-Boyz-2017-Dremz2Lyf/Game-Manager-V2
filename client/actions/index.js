// dependancies
import axios from 'axios';

// Action Types
import { SET_ERRORED, SET_LOADING, RECEIVE } from './types';

// Utilities
import { normalize } from '../utilities';

export const setErrored = (stateKey, data) => {
  // NOTE: Still not 100% sure if this is going to work for all errors.... but it might! :D
  let error = data.status ? {status: data.status, message:Â data.data} : {message: data.message, source: data.stack};

  return {
    type: SET_ERRORED,
    payload: {
      data: error,
      stateKey
    }
  };
};


export const setLoading = (stateKey, data) => {
  return {
    type: SET_LOADING,
    payload: {
      data,
      stateKey
    }
  };
};

export const receive = (stateKey, data) => {
  // receive is set up to accept an array or an object.
    // If it is an array we do the normalization here.
  if (Array.isArray(data)) {
    data = normalize(data);
  }

  // and send off our action with the proper payload
  return {
    type: RECEIVE,
    payload: {
      data,
      stateKey
    }
  };
};

export const fetch = (stateKey, data = {}) => {
  let request = axios.get(`/${stateKey}`, {params: data});

  return dispatch => {
    dispatch(setLoading(stateKey, true));

    return request.then(response => response.data).then(data => {
      dispatch(receive(stateKey, data));
    }).catch(error => {
      // This should handle any response errors, but more than likely not anything else.... D:
      let data = error.response ? error.response : error;
      dispatch(setErrored(stateKey, data));
    }).then(dis => {
      dispatch(setLoading(stateKey, false));
    });
  };
};

// In an effort to minimize request traffic, the server will respond with the updated or created data object we are currently working with. Instead of fetching all of that corresponding data we need after the post. We will just pass the updated data through to the receive action, by using getState from redux-thunk.

export const create = (stateKey, data) => {
  let post = axios.post(`/${stateKey}`, data);

  return (dispatch, getState) => {
    dispatch(setLoading(stateKey, true));

    return post.then(response => normalize(response.data)).then(updated => {

      // Grab the specific subState for the stateKey we are working with during this action
      let subState = getState().data[stateKey];

      // Take the subState and the updated data and create a new object
      let updatedSubState = {...subState, ...updated};

      // Pass that updated subState into the receive action
      return dispatch(receive(stateKey, updatedSubState));
    }).catch(error => {

      // If there was an error in the put, we grab the response from the server and pass it along to our setErrored action creator.
      let data = error.response;
      return dispatch(setErrored(stateKey, data));
    }).then(dis =>

      // Once everything is done, we set our loading flag to false.
      dispatch(setLoading(stateKey, false))
    );
  };
};

export const update = (stateKey, data) => {
  let put = axios.put(`/${stateKey}`, data);

  return (dispatch, getState) => {
    // Set our loading flag for the specific stateKey we are working with
    dispatch(setLoading(stateKey, true));

    // Then handle the async request.
    return put.then(response => normalize(response.data)).then(updated => {

      // Grab the specific subState for the stateKey we are working with during this action
      let subState = getState().data[stateKey];

      // Take the subState and the updated data and create a new object
      let updatedSubState = {...subState, ...updated};

      // Pass that updated subState into the receive action
      return dispatch(receive(stateKey, updatedSubState));
    }).catch(error => {

      // If there was an error in the put, we grab the response from the server and pass it along to our setErrored action creator.

      let data = error.response;
      return dispatch(setErrored(stateKey, data));
    }).then(dis =>

      // Once everything is done, we set our loading flag to false.
      dispatch(setLoading(stateKey, false))
    );
  };
};
