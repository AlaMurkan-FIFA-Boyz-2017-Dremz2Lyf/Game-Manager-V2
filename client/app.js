// Dependancies
import React from 'react';
import ReactDOM from 'react-dom';
import { ProviderÂ } from 'react-redux';
import { createStore, applyMiddleware, compose } from 'redux';
import promise from 'redux-promise';

// Components
import Home from './components/home';

import rootReducer from './reducers/index.js';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(rootReducer, undefined, composeEnhancers(
  applyMiddleware(promise)
));

ReactDOM.render(
  <Provider store={store}>
    <Home/>
  </Provider>,
  document.getElementById('app')
);
