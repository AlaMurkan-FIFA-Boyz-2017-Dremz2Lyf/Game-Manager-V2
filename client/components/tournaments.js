import React, { Component } from 'react';
import { Panel, Nav, NavItem } from 'react-bootstrap';
import { connect } from 'react-redux';

class Tournaments extends Component {

  handleSelect(eventKey) {
    console.log(eventKey);

  }

  render() {
    let activeKey = 'onGoing';
    return (
      <div className="tournaments">
        <Panel>
          <Nav bsStyle="tabs" activeKey={activeKey} onSelect={this.handleSelect}>
            <NavItem eventKey="finished">Finished</NavItem>
            <NavItem eventKey="onGoing">OnGoing</NavItem>
            <NavItem eventKey="create">Create</NavItem>
          </Nav>
        </Panel>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    activeKey: state.tournaments.activeKey,
    onGoing: state.tournaments.onGoing,
    finished: state.tournaments.finished
  };
};

export default connect(mapStateToProps, null)(Tournaments);
