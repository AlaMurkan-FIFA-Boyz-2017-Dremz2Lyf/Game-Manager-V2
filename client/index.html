<html>
<head>
  <title>Game Manager</title>
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans|Oswald|PT+Sans">
</head>
<body>
  <div id="app"></div>
</body>
</html>
