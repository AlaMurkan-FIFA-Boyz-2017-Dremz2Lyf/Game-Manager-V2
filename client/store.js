
// Make our store exportable for using in tests/and production

import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';

import rootReducer from './reducers/index.js';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(rootReducer, undefined, composeEnhancers(
  applyMiddleware(thunk)
));
