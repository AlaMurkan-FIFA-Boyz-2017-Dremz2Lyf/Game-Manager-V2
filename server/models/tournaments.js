const model = require('./lib/knex_model');

let tournaments = model.create('tournaments', {
  tableName: 'tournaments'
});

module.exports = tournaments;
