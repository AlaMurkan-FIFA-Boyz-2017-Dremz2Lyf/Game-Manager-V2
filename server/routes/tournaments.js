const express = require('express');
const router = express.Router();
const tournaments = require('../models/tournaments');
const games = require('../models/games');
const _ = require('../utilities');

router.get('/', (req, res) => {
  if (req.query.id) {
    tournaments.find(req.query.id).then(tournament => {
      res.send(tournament);
    }).catch(err => {
      res.status(404).send(err);
    });
  } else {
    tournaments.all().then(tournaments => {
      res.send(tournaments);
    }).catch(err => {
      res.status(404).send(err);
    });
  }
});

router.post('/', (req, res) => {
  // Check if there was no body posted send an error message if so.
  if (!req.body) {
    res.status(400).send({message: 'Whoops! Did you forget to include any info?'});
  }

  // Here we pull the necessary stuff from the request body
  let { added, rounds = 1, name } = req.body;

  tournaments.create({name}).then(tournament => (
    tournament
  )).then(tournament => {

    let newGames = _.createGames(added.map(Number), tournament.id, rounds);
    games.create(newGames).then(games => {
      let totalGames = Object.keys(games).length - 1;
      let { id } = tournament;
      tournaments.save({id, totalGames}).then(tournament => {
        res.status(201).send(tournament);
      }).catch(err => {
        // Catch for saving tournament with totalGames count
        res.status(500).send(err);
      });
    }).catch(err => {
      // Catch for creating the games in the db with the tournament id
      res.status(500).send(err);
    });
  }).catch(err => {
    // Catch for creating the tournament with just the name
    res.status(500).send(err);
  });
});

router.put('/', (req, res) => {
  let tournament = req.body;

  tournaments.updateOne(tournament).then(updatedTournament => {
    res.status(202).send(updatedTournament);
  }).catch(err => {
    if (err.type === 'invalid_argument') {
      res.status(400).send(err.type);
    } else {
      res.status(500).send(err);
    }
  });

});

module.exports = router;
