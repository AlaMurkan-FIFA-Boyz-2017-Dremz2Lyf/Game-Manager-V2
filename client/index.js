// Dependancies
import React from 'react';
import ReactDOM from 'react-dom';
import { ProviderÂ } from 'react-redux';

// css
import '../styles/styles.css';

// Bring in our router
import Routes from './router';

// Grab our root reducer and our store from their files
import rootReducer from './reducers/index.js';
import { store } from './store';

// Found this... Needed for hot module reloading.
if (module.hot) {
  module.hot.accept();
}

ReactDOM.render(
  <Provider store={store}>
    <Routes />
  </Provider>,
  document.getElementById('app')
);
