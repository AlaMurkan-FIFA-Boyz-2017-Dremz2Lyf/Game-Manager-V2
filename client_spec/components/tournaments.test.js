// Dependancies
import React from 'react';
import { ProviderÂ } from 'react-redux';

// Components
import { Tournaments, mapStateToProps } from '../../client/components/tournaments.js';

// Grab our store and our action to prePopulate
import { store } from '../../client/store';

describe('Tournaments Component', () => {
  let wrapper = shallow(<Tournaments tournaments={mockData.tournaments}/>);
  let panel = wrapper.find('Panel');

  test('should match snapshot', () => {

    expect(wrapper).toMatchSnapshot();
  });

  test('should render a tournament element', () => {

    expect(wrapper.is('.tournaments')).toBe(true);
  });

  test('should default to onGoing', () => {

    expect(wrapper.contains(<h4>OnGoing Tournaments</h4>)).toBe(true);
  });

  test('should call componentDidMount', () => {
    let mockFetch = jest.fn();
    let mounted = mount(
      <Provider store={store}>
        <Tournaments fetch={mockFetch} tournaments={mockData.tournaments}/>
      </Provider>
    );

    expect(mockFetch.mock.calls[0]).toEqual(['tournaments']);
  });

  test('should have a mapStateToProps', () => {
    let fakeStore = {
      data: {
        tournaments: {}
      }
    };

    expect(mapStateToProps(fakeStore)).toEqual({tournaments: {}});
  });

  test('render list should build a list based on the boolean it was called with', () => {
    let finished = wrapper.instance().renderList(true);
    let onGoing = wrapper.instance().renderList(false);

    expect(finished).toHaveLength(1);
    expect(onGoing).toHaveLength(4);
  });

  test('should handle no props passed to it', () => {
    let empty = shallow(<Tournaments/>);

    expect(empty).toMatchSnapshot();

  });

  describe('Panel Child Component', () => {

    test('should exist', () => {

      let count = panel.length;

      expect(count).toBe(1);
    });

    test('should have three Tabs', () => {

      expect(wrapper.find('Tab')).toHaveLength(3);
    });

    test('Tabs should have titles of "onGoing", "finished", & "create"', () => {
      expect(wrapper.find('[title="OnGoing"]')).toHaveLength(1);
      expect(wrapper.find('[title="Finished"]')).toHaveLength(1);
      expect(wrapper.find('[title="Create One"]')).toHaveLength(1);
    });

    test('should render different content for the correct tab', () => {

      wrapper.find('Tab').find('[title="Finished"]').simulate('click');
      expect(wrapper.contains(<h4>Finished Tournaments</h4>)).toBe(true);
    });

  });

});
