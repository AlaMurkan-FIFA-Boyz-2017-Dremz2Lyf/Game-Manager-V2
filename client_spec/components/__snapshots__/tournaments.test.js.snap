exports[`Tournaments Component should handle no props passed to it 1`] = `
<Carousel
  activeIndex={1}
  bsClass="carousel"
  controls={true}
  direction={null}
  indicators={true}
  interval={5000}
  nextIcon={
    <Glyphicon
      bsClass="glyphicon"
      glyph="chevron-right" />
  }
  nextLabel="Next"
  onSelect={[Function]}
  pauseOnHover={true}
  prevIcon={
    <Glyphicon
      bsClass="glyphicon"
      glyph="chevron-left" />
  }
  prevLabel="Previous"
  slide={true}
  wrap={true}>
  <CarouselItem
    active={false}
    animateIn={false}
    animateOut={false}>
    <ListGroup
      bsClass="list-group"
      className="tournament-list" />
    <CarouselCaption
      bsClass="carousel-caption"
      componentClass="div">
      <h3>
        Finished Tournaments
      </h3>
    </CarouselCaption>
  </CarouselItem>
  <CarouselItem
    active={false}
    animateIn={false}
    animateOut={false}>
    <ListGroup
      bsClass="list-group"
      className="tournament-list" />
    <CarouselCaption
      bsClass="carousel-caption"
      componentClass="div">
      <h3>
        OnGoing Tournaments
      </h3>
    </CarouselCaption>
  </CarouselItem>
  <CarouselItem
    active={false}
    animateIn={false}
    animateOut={false}>
    <Connect(CreateTournament) />
  </CarouselItem>
</Carousel>
`;

exports[`Tournaments Component should match snapshot 1`] = `
<Carousel
  activeIndex={1}
  bsClass="carousel"
  controls={true}
  direction={null}
  indicators={true}
  interval={5000}
  nextIcon={
    <Glyphicon
      bsClass="glyphicon"
      glyph="chevron-right" />
  }
  nextLabel="Next"
  onSelect={[Function]}
  pauseOnHover={true}
  prevIcon={
    <Glyphicon
      bsClass="glyphicon"
      glyph="chevron-left" />
  }
  prevLabel="Previous"
  slide={true}
  wrap={true}>
  <CarouselItem
    active={false}
    animateIn={false}
    animateOut={false}>
    <ListGroup
      bsClass="list-group"
      className="tournament-list">
      <Tournament
        tournament={
          Object {
            "createdAt": "2016-12-25T19:31:48",
            "gamesPlayed": 1,
            "id": 2,
            "name": "new",
            "totalGames": 1,
            "updatedAt": "2016-12-26T19:31:48",
            "winner": 2,
          }
        } />
    </ListGroup>
    <CarouselCaption
      bsClass="carousel-caption"
      componentClass="div">
      <h3>
        Finished Tournaments
      </h3>
    </CarouselCaption>
  </CarouselItem>
  <CarouselItem
    active={false}
    animateIn={false}
    animateOut={false}>
    <ListGroup
      bsClass="list-group"
      className="tournament-list">
      <Tournament
        tournament={
          Object {
            "createdAt": "2016-12-25T19:31:48",
            "gamesPlayed": 0,
            "id": 1,
            "name": "Super Tourney!",
            "totalGames": 3,
            "updatedAt": null,
            "winner": null,
          }
        } />
      <Tournament
        tournament={
          Object {
            "createdAt": "2016-12-25T19:31:48",
            "gamesPlayed": 5,
            "id": 3,
            "name": "next",
            "totalGames": 5,
            "updatedAt": null,
            "winner": null,
          }
        } />
      <Tournament
        tournament={
          Object {
            "createdAt": "2016-12-25T19:31:48",
            "gamesPlayed": 1,
            "id": 4,
            "name": "hot",
            "totalGames": 7,
            "updatedAt": null,
            "winner": null,
          }
        } />
      <Tournament
        tournament={
          Object {
            "createdAt": "2016-12-25T19:31:48",
            "gamesPlayed": 2,
            "id": 5,
            "name": "fun",
            "totalGames": 6,
            "updatedAt": null,
            "winner": null,
          }
        } />
    </ListGroup>
    <CarouselCaption
      bsClass="carousel-caption"
      componentClass="div">
      <h3>
        OnGoing Tournaments
      </h3>
    </CarouselCaption>
  </CarouselItem>
  <CarouselItem
    active={false}
    animateIn={false}
    animateOut={false}>
    <Connect(CreateTournament) />
  </CarouselItem>
</Carousel>
`;
